lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -0,0 +1,247 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	BlockAffinityResourceName = "BlockAffinities"
+	BlockAffinityCRDName      = "blockaffinities.crd.projectcalico.org"
+)
+
+func NewAffinityBlockClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
+	// Create a resource client which manages k8s CRDs.
+	rc := customK8sResourceClient{
+		clientSet:       c,
+		restClient:      r,
+		name:            BlockAffinityCRDName,
+		resource:        BlockAffinityResourceName,
+		description:     "Calico IPAM block affinities",
+		k8sResourceType: reflect.TypeOf(apiv3.BlockAffinity{}),
+		k8sResourceTypeMeta: metav1.TypeMeta{
+			Kind:       apiv3.KindBlockAffinity,
+			APIVersion: apiv3.GroupVersionCurrent,
+		},
+		k8sListType:  reflect.TypeOf(apiv3.BlockAffinityList{}),
+		resourceKind: apiv3.KindBlockAffinity,
+	}
+
+	return &affinityBlockClient{rc: rc}
+}
+
+// Implements the api.Client interface for AffinityBlocks.
+type affinityBlockClient struct {
+	rc customK8sResourceClient
+}
+
+func (c affinityBlockClient) toV1(kvpv3 *model.KVPair) (*model.KVPair, error) {
+	cidrStr := kvpv3.Value.(*apiv3.BlockAffinity).Annotations["projectcalico.org/cidr"]
+	host := kvpv3.Value.(*apiv3.BlockAffinity).Annotations["projectcalico.org/host"]
+	_, cidr, err := net.ParseCIDR(cidrStr)
+	if err != nil {
+		log.WithField("cidr", cidr).WithError(err).Error("failed to parse cidr")
+		return nil, err
+	}
+	state := model.BlockAffinityState(kvpv3.Value.(*apiv3.BlockAffinity).Spec.State)
+	return &model.KVPair{
+		Key: model.BlockAffinityKey{
+			Host: host,
+			CIDR: *cidr,
+			UID:  &kvpv3.Value.(*apiv3.BlockAffinity).UID,
+		},
+		Value: &model.BlockAffinity{
+			State: state,
+		},
+		Revision: kvpv3.Revision,
+	}, nil
+}
+
+func (c affinityBlockClient) v3Fields(k model.Key) (name, cidr, host string) {
+	// Sanitize the CIDR, replacing characters which
+	// are not allowed in the Kubernetes API.
+	// e.g., 10.0.0.1/26 -> 10-0-0-1-26
+	host = k.(model.BlockAffinityKey).Host
+	cidr = fmt.Sprintf("%s", k.(model.BlockAffinityKey).CIDR)
+	cidrstr := strings.Replace(cidr, ".", "-", -1)
+	cidrstr = strings.Replace(cidrstr, ":", "-", -1)
+	cidrstr = strings.Replace(cidrstr, "/", "-", -1)
+
+	// Include the hostname as well.
+	host = k.(model.BlockAffinityKey).Host
+	name = fmt.Sprintf("%s-%s", host, cidrstr)
+
+	if len(name) >= 253 {
+		// If the name is too long, we need to shorten it.
+		// Remove enough characters to get it below the 253 character limit,
+		// as well as 11 characters to add a hash which helps with uniqueness,
+		// and two characters for the `-` separators between clauses.
+		name = fmt.Sprintf("%s-%s", host[:252-len(cidrstr)-13], cidrstr)
+
+		// Add a hash to help with uniqueness.
+		// Kubernetes requires all names to end with an alphabetic character, so
+		// append a 'c' to the end to ensure we always meet this requirement.
+		h := sha1.New()
```

Comment:
sha1 is known-broken; best to use sha-256 if only because it means that grepping the codebase for sha1 will turn up no hits, making it easier to know that we're not using it where we shouldn't.  (Unlikely that it's breakage affects this use case!)